rfd/0050-join-methods.md

```diff
@@ -0,0 +1,118 @@
+
+#### Auth `POST /tokens/register/auth`
+Join Methods: token
+Server trust: CA pins
+Looks like this is a legacy endpoint the just deletes the given token. Only
+works for tokens with the `Auth` role. I can't find anything that calls this, so
+it can probably be scheduled for deletion.
+
+#### Auth `rpc GenerateHostCerts(HostCertsRequest) returns (Certs)`
+Client trust: mTLS
+Server trust: mTLS
+This is an mTLS authenticated gRPC endpoint for cert renewal. The node must
+first join the cluster using another endpoint to get its first client
+certificate used for the mTLS connection.
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes that don't have a direct connection to the auth server, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### IAM Join Method
+The IAM join method (introduced in RFD 41) requires gRPC methods rather than
+HTTP because the design requires streams to implement a challenge/response
+protocol. To summarize, after the client initiates the rpc the Auth server sends
+a crypto-random "challenge" string over the server->client stream. This
+challenge must be included in a signed `sts:GetCallerIdentity` request which
+will be sent on the client->server stream.  Finally, if everything checks out,
+the Auth server will send the signed certificates on the server->client stream.
+
+A new gRPC streaming method will be added on Auth and Proxy to support this. The
+client will be able to call the gRPC method with either a Proxy or Auth address
+and it will "just work" transparently.
+
+#### Auth `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join methods: IAM
+Server trust: CA pins
+This is the Auth endpoint that will complete the IAM join request.
+
+#### Proxy `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join methods: IAM
+Server trust: PKI
+Normally, authenticated gRPC calls from IoT nodes are tunnelled through the
+proxy over SSH. This will not work for unauthenticated clients which don't yet
+have an SSH certificate.
+
+It would be possible to use TLS routing on the proxy to forward gRPC requests
+directly to the Auth server, but this would open up the Auth server to DOS
+attacks from unauthenticated clients.
+
+Instead, this single gRPC method will be implemented on the proxy. This method
+will first perform rate-limiting, then "forward" the request to the auth server
+method at the application layer by calling the gRPC method through the Proxy's
+own Auth client.
+
+Since the Proxy currently does not expose any gRPC service of its own, a new
+gRPC listener will be added to support this. gRPC connections will be
+multiplexed on the web listener/port using the existing ALPN-based multiplexer
+and a special ALPN ProtocolName of `teleport-proxy-grpc` that will be passed by
```

Comment:
Using a custom ALPN value is not going to work, gRPC has to actually negotiate `h2` for HTTP/2. Could either use SNI, or demux based on the `Content-type: application/grpc` header in the root HTTP handler

Reply:
SNI seems to be the best, as we may elect to host/proxy different gRPC services.

Reply:
I figured out how to make ALPN work, we can route based on a custom value in the clients list of supported NextProtos, but then actually negotiate `h2`